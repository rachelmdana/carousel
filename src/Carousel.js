import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";


/** Carousel: displays images and arrows to navigate through them
 * 
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 * 
 * State:
 * - currCardIdx: integer for current card index
 * 
 * App --> Carousel --> Card
 */
function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  const currCard = photos[currCardIdx];
  const total = photos.length;

  function goForward(direction) {
    if (direction === 'left') {
      if (currCardIdx > 0) {
        setCurrCardIdx(currCardIdx - 1);
      }
    } else if (direction === 'right') {
      if (currCardIdx < total - 1) {
        setCurrCardIdx(currCardIdx + 1);
      }
    }
  }

  return (
    <div className="Carousel">
      <h1>{title}</h1>
     <div className="Carousel-main">
        <i
          className="bi bi-arrow-left-circle"
          onClick={() => goForward('left')}
          data-testid="leftArrow"
          disabled={currCardIdx === 0}
        />
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        <i
          className="bi bi-arrow-right-circle"
          onClick={() => goForward('right')}
          data-testid="rightArrow"
          disabled={currCardIdx === total - 1}
        />
      </div>
    </div>
  );
}

export default Carousel;
