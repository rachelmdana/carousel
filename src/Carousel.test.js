import React from "react";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import Carousel from "./Carousel";
import photos from "./photos.js";
// import '@testing-library/jest-dom/extend-expect';
import TEST_IMAGES from "./_testCommon.js";


it('renders Carousel without crashing', () => {
  render(<Carousel photos={photos} title="Test Carousel" />);
});

test('renders Carousel correctly', () => {
  const { asFragment } = render(<Carousel photos={photos} title="Test Carousel" />);
  expect(asFragment()).toMatchSnapshot();
});

it('left arrow moves to the previous image', () => {
  const photos = [
    { src: 'image1.jpg', caption: 'Image 1' },
    { src: 'image2.jpg', caption: 'Image 2' },
    { src: 'image3.jpg', caption: 'Image 3' },
  ];

  const { getByTestId } = render(<Carousel photos={photos} title="Test Carousel" />);


  // Initially on the first image (index 0)
  expect(getByTestId('Card-title').textContent).toBe('Image 1');

  // Click the left arrow to move to the second image (index 1)
  fireEvent.click(getByTestId('leftArrow'));

  // Now test expects to be on the second image
  expect(getByTestId('Card-title').textContent).toBe('Image 2');
});

it('left arrow is missing on the first image and right arrow is missing on the last image', () => {
  const photos = [
    { src: 'image1.jpg', caption: 'Image 1' },
    { src: 'image2.jpg', caption: 'Image 2' },
    { src: 'image3.jpg', caption: 'Image 3' },
  ];

  const { getByTestId } = render(<Carousel photos={photos} title="Test Carousel" />);

  // Initially on the first image
  expect(getByTestId('leftArrow')).toHaveAttribute('disabled');
  expect(getByTestId('rightArrow')).not.toHaveAttribute('disabled');

  // Click the right arrow to move to the second image
  fireEvent.click(getByTestId('rightArrow'));

  // Now on the second image
  expect(getByTestId('leftArrow')).not.toHaveAttribute('disabled');
  expect(getByTestId('rightArrow')).not.toHaveAttribute('disabled');

  // Click the right arrow to move to the third image
  fireEvent.click(getByTestId('rightArrow'));

  // Now on the third image
  expect(getByTestId('leftArrow')).not.toHaveAttribute('disabled');
  expect(getByTestId('rightArrow')).toHaveAttribute('disabled');
});

it("works when you click on the right arrow", function() {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).not.toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument();
});
