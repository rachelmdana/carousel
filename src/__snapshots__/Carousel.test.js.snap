// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Carousel correctly 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Test Carousel
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
        data-testid="leftArrow"
        disabled=""
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
          data-testid="Card-title"
        >
          Photo by Richard Pasquarella on Unsplash
        </h4>
        <img
          alt="Photo by Richard Pasquarella on Unsplash"
          class="Card-image"
          src="image1.jpg"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        data-testid="rightArrow"
      />
    </div>
  </div>
</DocumentFragment>
`;
